esp/esp/program/controllers/autoscheduler/db_interface.py

```diff
@@ -0,0 +1,757 @@
+def load_resource_constraints(
+        program, specification_overrides=None, specs_only=False,
+        ignore_comments=True):
+    if specification_overrides is None:
+        specification_overrides = {}
+
+    tag_value = Tag.getProgramTag(config.RESOURCE_CONSTRAINTS_TAG,
+                                  program=program,
+                                  default="{}")
+    try:
+        tag_overrides = json.loads(tag_value)
+        if ignore_comments:
+            tag_overrides = {
+                k: v for k, v in tag_overrides.iteritems()
+                if "_comment" not in k}
+    except ValueError as e:
+        raise SchedulingError(
+            "Resource constraints Tag is malformatted with error {}: {}"
+            .format(e, tag_value))
+    specs = [config.DEFAULT_RESOURCE_CONSTRAINTS,
+             tag_overrides,
+             specification_overrides]
+    if specs_only:
+        return {
+            name: spec for name, spec
+            in util.override(specs).iteritems()
+            if spec != "None" and spec is not None}
+    else:
+        valid_res_types = \
+            ResourceType.objects.filter(program=program).values_list(
+                    "name", flat=True)
+        return resource_checker.create_resource_criteria(
+                specs, valid_res_types)
+
+
+@util.timed_func("db_interface_load_resoure_scoring")
+def load_resource_scoring(
+        program, specification_overrides=None, specs_only=True,
+        ignore_comments=True):
+    if specification_overrides is None:
+        specification_overrides = {}
+
+    tag_value = Tag.getProgramTag(config.RESOURCE_SCORING_TAG,
+                                  program=program,
+                                  default="{}")
+    try:
+        tag_overrides = json.loads(tag_value)
+        if ignore_comments:
+            tag_overrides = {
+                k: v for k, v in tag_overrides.iteritems()
+                if "_comment" not in k}
+    except ValueError as e:
+        raise SchedulingError(
+            "Resource scoring Tag is malformatted with error {}: {}"
+            .format(e, tag_value))
+
+    specs = [config.DEFAULT_RESOURCE_SCORING,
+             tag_overrides,
+             specification_overrides]
+    if specs_only:
+        return {
+            name: (spec, weight) for name, (spec, weight)
+            in util.override(specs).iteritems()
+            if spec != "None" and spec is not None}
+    else:
+        valid_res_types = \
+            ResourceType.objects.filter(program=program).values_list(
+                    "name", flat=True)
+        return resource_checker.create_resource_criteria(
+                specs, valid_res_types, use_weights=True)
+
+
+@util.timed_func("db_interface_batch_convert_sections")
+def batch_convert_sections(
+        sections, program, teachers_by_section, timeslot_dict, rooms,
+        rooms_by_section, meeting_times_by_section,
+        requests_by_section):
+    return [s for s in
```

Comment:
Rewritten as a for loop, but I'm keeping it non-inlined so that it can be timed (although ideally the timer would have a context manager version instead of just a decorator version, lol). (iirc this is in fact on the critical path for loading.)